packages/mongodb-ts-autocomplete/src/index.ts

```diff
@@ -0,0 +1,279 @@
+  }
+
+  getConnectionCode(connectionKey: string): string {
+    return `
+/// <reference path="/shell-api.d.ts" />
+
+export {};
+
+declare global {
+  namespace Connection${connectionKey} {
+    export type ServerSchema = ${this.connectionSchemas[
+      connectionKey
+    ].toTypescriptTypeDefinition()}
+  }
+}
+`;
+  }
+
+  getCurrentGlobalsCode() {
+    if (!this.currentConnectionKey) {
+      throw new Error('No active connection');
+    }
+
+    const databaseName = this.getActiveConnection().getCurrentDatabaseName();
+    return `
+/// <reference path="/shell-api.d.ts" />
+/// <reference path="/${this.currentConnectionKey}.d.ts" />
+
+export {}; // turns this into an "external module"
+
+declare global {
+  type CurrentDatabaseSchema = Connection${this.currentConnectionKey}.ServerSchema['${databaseName}'];
+
+  var db: CurrentDatabaseSchema;
+  var use: (collection: string) => CurrentDatabaseSchema;
+}
+`;
+  }
+
+  async autocomplete(
+    code: string,
+    position?: number
+  ): Promise<AutoCompletion[]> {
+    if (!this.currentConnectionKey) {
+      throw new Error('No active connection');
+    }
+
+    if (typeof position === 'undefined') {
+      position = code.length;
+    }
+
+    // TODO: we're now compiling the source twice: autocomplete will also
+    // compile if it finds completions in order to filter them. Does it matter?
+    // Maybe. Not so much for a typical short shell line, probably more for a
+    // large document or aggregation.
+    const tsAst = compileSourceFile(code);
+    const collectionName = inferCollectionNameFromFunctionCall(tsAst) || 'test';
+
+    const mdbCmd = inferMongoDBCommandFromFunctionCall(tsAst);
+
+    let schema: JSONSchema;
+
+    if (
+      isInAggregationPipelinePosition(tsAst, position) ||
+      mdbCmd === 'aggregate'
+    ) {
+      const pipelineToDryRun =
+        extractPipelineUptoCaret(tsAst, position) ||
+        extractPipelineFromLastAggregate(tsAst, position) ||
+        [];
+
+      schema = await this.context.schemaInformationForAggregation(
+        this.currentConnectionKey,
+        this.getActiveConnection().getCurrentDatabaseName(),
+        collectionName,
+        pipelineToDryRun as Pipeline
+      );
+    } else {
+      schema = await this.context.schemaInformationForCollection(
+        this.currentConnectionKey,
+        this.getActiveConnection().getCurrentDatabaseName(),
+        collectionName
+      );
+    }
+
+    const activeConnection = this.getActiveConnection();
+    const databaseName = this.getActiveConnection().getCurrentDatabaseName();
+    activeConnection.addCollectionSchema(databaseName, collectionName, schema);
+
+    const collectionNames = await this.context.collectionsForDatabase(
+      this.currentConnectionKey,
+      databaseName
+    );
+    activeConnection.setDatabaseCollectionNames(databaseName, collectionNames);
+
+    // TODO: the problem with doing it this way is that, while the collection
```

Comment:
I think this is probably fine. Getting the sample docs and calculating a schema is what's relatively expensive, probably not walking it and writing a few lines of typescript. Only way this becomes an issue is if you have a really large collection schema or very many databases/collections in the same server schema. Each individual database's list of collections is lazy loaded, though and also each collection's schema. So you'd likely have to have quite the long lived and meandering session for that to build up.

We would likely load the list of databases or collections for at least one database early on, though.

Reply:
Similar to comments above, I think I'll remove this comment before merging. Just wanted to point out that I'm aware.

Reply:
I filed a follow-up ticket for this https://jira.mongodb.org/browse/MONGOSH-2209